### PR TITLE
Page merger strategy

### PR DESCRIPTION
**Pull request by sswistun-vl**
Implemented strategies for page merger.

### CODE REVIEW

File: core/src/main/kotlin/transformers/psi/DefaultPsiToDocumentationTranslator.kt [hunk 1]
```diff
@@ -100,14 +111,15 @@ object DefaultPsiToDocumentationTranslator : PsiToDocumentationTranslator {
...
+                getComment(psi),
+                visibility = mapOf(platformData to psi.getVisibility())
             )
         }
 
         private fun parseField(psi: PsiField, parent: DRI): Property {
             val dri = parent.copy(
                 callable = Callable(
-                    psi.name,
+                    psi.name ?: "",
```
Comment (kamilok1965) [inline core/src/main/kotlin/transformers/psi/DefaultPsiToDocumentationTranslator.kt]: I think `psi.name` returns `String`, so the elvis is redundant

File: gradle.properties [hunk 2]
```diff
@@ -6,6 +6,7 @@ kotlin_version=1.3.61
 kotlin_plugin_version=1.3.61-release-180
 idea_version=192.5728.98
 language_version=1.3
+org.gradle.jvmargs=-Xmx2g
```
Comment (kamilok1965) [inline gradle.properties]: Do we need this? If so, it should be at the bottom with a new comment

File: plugins/kotlin-as-java/build.gradle.txt [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import javax.tools.ToolProvider
```
Comment (kamilok1965) [inline plugins/kotlin-as-java/build.gradle.txt]: What's the purpose of this file?

File: plugins/kotlin-as-java/src/main/kotlin/KotlinAsJavaPageContentBuilder.kt [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+class KotlinAsJavaPageContentBuilder(
+    private val dri: Set<DRI>,
+    private val platformData: Set<PlatformData>,
+    private val kind: Kind,
+    private val commentsConverter: CommentsToContentConverter,
+    val logger: DokkaLogger,
+    private val styles: Set<Style> = emptySet(),
+    private val extras: Set<Extra> = emptySet()
+) : PageContentBuilder {
```
Comment (kamilok1965) [inline plugins/kotlin-as-java/src/main/kotlin/KotlinAsJavaPageContentBuilder.kt]: Subclassing `DefaultPageContentBuilder` is a good idea there IMHO, as some methods are just duplicates

File: settings.gradle.kts [hunk 5]
```diff
@@ -8,8 +8,8 @@ include("runners:cli")
 include("runners:maven-plugin")
 include("plugins:xml")
 include("plugins:mathjax")
-include("integration-tests:gradle-integration-tests")
-
+include("plugins:kotlin-as-java")
+//include("integration-tests:gradle-integration-tests")
```
Comment (kamilok1965) [inline settings.gradle.kts]: Don't comment it, if you want to skip tests better to uncomment a line in integration-tests' build.gradle

File: core/src/main/kotlin/transformers/pages/PageNodeTransformer.kt [hunk 6]
```diff
@@ -4,6 +4,6 @@ import org.jetbrains.dokka.pages.ModulePageNode
 import org.jetbrains.dokka.pages.RootPageNode
 import org.jetbrains.dokka.plugability.DokkaContext
 
-interface PageNodeTransformer {
-    operator fun invoke(input: RootPageNode): RootPageNode
+abstract class PageNodeTransformer(context: DokkaContext? = null) {
+    abstract operator fun invoke(input: RootPageNode): RootPageNode
 }
```
Comment (kamilok1965) [inline core/src/main/kotlin/transformers/pages/PageNodeTransformer.kt]: This change is probably unnecessary

File: core/src/main/kotlin/renderers/html/htmlPreprocessors.kt [hunk 7]
```diff
@@ -43,20 +43,20 @@ object NavigationPageInstaller : PageNodeTransformer {
 
     private fun visit(page: ContentPage): NavigationNode = NavigationNode(
         page.name,
-        page.dri,
+        page.dri.first(),
```
[the remark this answers is not in the document]
Comment (Kordyjan) [reply core/src/main/kotlin/renderers/html/htmlPreprocessors.kt]: It works as long as there is one-to-many relation between pages and dris. If we have many-to-one or many-to-many, our whole location resolution breaks down, so it right here is not the biggest problem.

File: core/src/main/kotlin/transformers/pages/PageMergerStrategy.kt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package org.jetbrains.dokka.transformers.pages
+
+import org.jetbrains.dokka.pages.PageNode
+
+abstract class PageMergerStrategy {
```
[the remark this answers is not in the document]
Comment (Kordyjan) [reply core/src/main/kotlin/transformers/pages/PageMergerStrategy.kt]: Moreover interfaces don't restrict adding any further supertypes while classes do. It is strictly better to use interface when possible.

File: core/src/main/kotlin/transformers/pages/PageNodeTransformer.kt (hunk 6 above)
Comment (Kordyjan) [reply core/src/main/kotlin/transformers/pages/PageNodeTransformer.kt]: Yup. Extension points' interfaces(!) shouldn't be aware of context. Only their implementation should. It is done this way in every other EP.

File: core/src/main/kotlin/transformers/pages/DefaultPageMergerStrategy.kt [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package org.jetbrains.dokka.transformers.pages
+
+import org.jetbrains.dokka.pages.*
+
+object DefaultPageMergerStrategy : PageMergerStrategy() {
```
Comment (Kordyjan) [inline core/src/main/kotlin/transformers/pages/DefaultPageMergerStrategy.kt]: The idea of this class was that it "merges" pages by just returning the first one and reporting warning if group contains more than one page.

Comment (sswistun-vl) [reply core/src/main/kotlin/transformers/pages/DefaultPageMergerStrategy.kt]: Now there are Default and SamePage strategies.

File: core/src/main/kotlin/pages/PageNodes.kt [hunk 10]
```diff
@@ -135,11 +138,12 @@ class MemberPageNode(
     override val embeddedResources: List<String> = listOf()
 ) : ContentPage {
     override fun modified(name: String, children: List<PageNode>): MemberPageNode =
-        modified(name = name, content = this.content, children = children)
+        modified(name = name, content = this.content, children = children) as MemberPageNode
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply core/src/main/kotlin/pages/PageNodes.kt]: Bump

**Pull request merged**